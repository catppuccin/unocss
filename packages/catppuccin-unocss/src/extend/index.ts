import {
  flavorEntries,
  flavors,
  type CatppuccinFlavor,
} from '@catppuccin/palette';

import type { ThemeExtender } from '@unocss/core';

import type { ExtendOptions } from './types.ts';

/**
 * Extend the `theme` object of UnoCSS by passing this
 * function to its `extendTheme` option.
 *
 * @param options - Options for the extender
 */
export const _extendTheme = (options: ExtendOptions = {}) => {
  const { themeKey = 'colors', prefix = 'ctp', defaultFlavour } = options;

  /**
   * Adds Catppuccin colours to the target object.
   *
   * @param targetObj - Target theme object to extend
   * @param flavour - The Catppuccin flavour object
   * @param namespace - Optional namespace to nest colours under
   */
  const addFlavourColours = (
    targetObj: ThemeColoursObject,
    flavour: CatppuccinFlavor,
    namespace?: string,
  ) => {
    const colourEntries = Object.entries(flavour.colors);
    let targetContainer: ThemeColoursObject;

    if (namespace) {
      if (
        targetObj[namespace] !== undefined
        && typeof targetObj[namespace] !== 'object'
      ) {
        if (!targetObj['ctp']) targetObj['ctp'] = {};
        const targetObjCtp = targetObj['ctp'] as ThemeColoursObject;

        if (!targetObjCtp[namespace]) targetObjCtp[namespace] = {};
        targetContainer = targetObjCtp[namespace] as ThemeColoursObject;
      } else {
        if (!targetObj[namespace]) targetObj[namespace] = {};
        targetContainer = targetObj[namespace];
      }
    } else if (prefix) {
      targetContainer = targetObj;
    } else {
      targetContainer = targetObj;
    }

    for (const [colourName, colourData] of colourEntries) {
      if (!prefix && colourName in targetObj) {
        if (!targetObj['ctp']) targetObj['ctp'] = {};
        const targetObjCtp = targetObj['ctp'] as ThemeColoursObject;

        targetObjCtp[colourName] = colourData.hex;
      } else {
        targetContainer[colourName] = colourData.hex;
      }
    }
  };

  return (baseTheme => {
    const theme = baseTheme as Record<string, ThemeColoursObject>;

    if (!theme[themeKey]) theme[themeKey] = {};

    let targetObject = theme[themeKey];
    if (prefix) {
      if (!targetObject[prefix]) targetObject[prefix] = {};
      targetObject = targetObject[prefix] = targetObject[
        prefix
      ] as ThemeColoursObject;
    }

    if (defaultFlavour && defaultFlavour in flavors) {
      addFlavourColours(targetObject, flavors[defaultFlavour]);
    } else {
      for (const [flavourIdentifier, flavour] of flavorEntries) {
        addFlavourColours(targetObject, flavour, flavourIdentifier);
      }
    }
  }) satisfies ThemeExtender;
};

/**
 * Nested object structure for theme colours.
 *
 * @internal
 */
export interface ThemeColoursObject {
  [key: string]: ThemeColoursObject | string;
}
