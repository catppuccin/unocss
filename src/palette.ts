/**
 * This file is auto-generated by `generate-palette`
 * npm script. Do NOT edit manually.
 */

export const FLAVOUR_NAMES = ['latte', 'frappe', 'macchiato', 'mocha'] as const;
export const FLAVOUR_NAMES_PRETTY = [
  'Latte',
  'FrappÃ©',
  'Macchiato',
  'Mocha',
] as const;
export const FLAVOUR_EMOJIS = ['ðŸŒ»', 'ðŸª´', 'ðŸŒº', 'ðŸŒ¿'] as const;

export const ACCENT_COLOUR_NAMES = [
  'rosewater',
  'flamingo',
  'pink',
  'mauve',
  'red',
  'maroon',
  'peach',
  'yellow',
  'green',
  'teal',
  'sky',
  'sapphire',
  'blue',
  'lavender',
] as const;
export const NEUTRAL_COLOUR_NAMES = [
  'text',
  'subtext1',
  'subtext0',
  'overlay2',
  'overlay1',
  'overlay0',
  'surface2',
  'surface1',
  'surface0',
  'base',
  'mantle',
  'crust',
] as const;

export type FlavourName = (typeof FLAVOUR_NAMES)[number];
export type PrettyFlavourName = (typeof FLAVOUR_NAMES_PRETTY)[number];
export type FlavourEmoji = (typeof FLAVOUR_EMOJIS)[number];

export type ColourName = AccentColourName | NeutralColourName;
export type AccentColourName = (typeof ACCENT_COLOUR_NAMES)[number];
export type NeutralColourName = (typeof NEUTRAL_COLOUR_NAMES)[number];

export type FlavourPalette = Record<ColourName, `#${string}`>;
export type ColourPalette = Record<FlavourName, FlavourPalette>;

export const COLOURS_FLAVOUR_LATTE = {
  [ACCENT_COLOUR_NAMES[0]]: '#dc8a78',
  [ACCENT_COLOUR_NAMES[1]]: '#dd7878',
  [ACCENT_COLOUR_NAMES[2]]: '#ea76cb',
  [ACCENT_COLOUR_NAMES[3]]: '#8839ef',
  [ACCENT_COLOUR_NAMES[4]]: '#d20f39',
  [ACCENT_COLOUR_NAMES[5]]: '#e64553',
  [ACCENT_COLOUR_NAMES[6]]: '#fe640b',
  [ACCENT_COLOUR_NAMES[7]]: '#df8e1d',
  [ACCENT_COLOUR_NAMES[8]]: '#40a02b',
  [ACCENT_COLOUR_NAMES[9]]: '#179299',
  [ACCENT_COLOUR_NAMES[10]]: '#04a5e5',
  [ACCENT_COLOUR_NAMES[11]]: '#209fb5',
  [ACCENT_COLOUR_NAMES[12]]: '#1e66f5',
  [ACCENT_COLOUR_NAMES[13]]: '#7287fd',
  [NEUTRAL_COLOUR_NAMES[0]]: '#4c4f69',
  [NEUTRAL_COLOUR_NAMES[1]]: '#5c5f77',
  [NEUTRAL_COLOUR_NAMES[2]]: '#6c6f85',
  [NEUTRAL_COLOUR_NAMES[3]]: '#7c7f93',
  [NEUTRAL_COLOUR_NAMES[4]]: '#8c8fa1',
  [NEUTRAL_COLOUR_NAMES[5]]: '#9ca0b0',
  [NEUTRAL_COLOUR_NAMES[6]]: '#acb0be',
  [NEUTRAL_COLOUR_NAMES[7]]: '#bcc0cc',
  [NEUTRAL_COLOUR_NAMES[8]]: '#ccd0da',
  [NEUTRAL_COLOUR_NAMES[9]]: '#eff1f5',
  [NEUTRAL_COLOUR_NAMES[10]]: '#e6e9ef',
  [NEUTRAL_COLOUR_NAMES[11]]: '#dce0e8',
} as const satisfies FlavourPalette;
export const COLOURS_FLAVOUR_FRAPPE = {
  [ACCENT_COLOUR_NAMES[0]]: '#f2d5cf',
  [ACCENT_COLOUR_NAMES[1]]: '#eebebe',
  [ACCENT_COLOUR_NAMES[2]]: '#f4b8e4',
  [ACCENT_COLOUR_NAMES[3]]: '#ca9ee6',
  [ACCENT_COLOUR_NAMES[4]]: '#e78284',
  [ACCENT_COLOUR_NAMES[5]]: '#ea999c',
  [ACCENT_COLOUR_NAMES[6]]: '#ef9f76',
  [ACCENT_COLOUR_NAMES[7]]: '#e5c890',
  [ACCENT_COLOUR_NAMES[8]]: '#a6d189',
  [ACCENT_COLOUR_NAMES[9]]: '#81c8be',
  [ACCENT_COLOUR_NAMES[10]]: '#99d1db',
  [ACCENT_COLOUR_NAMES[11]]: '#85c1dc',
  [ACCENT_COLOUR_NAMES[12]]: '#8caaee',
  [ACCENT_COLOUR_NAMES[13]]: '#babbf1',
  [NEUTRAL_COLOUR_NAMES[0]]: '#c6d0f5',
  [NEUTRAL_COLOUR_NAMES[1]]: '#b5bfe2',
  [NEUTRAL_COLOUR_NAMES[2]]: '#a5adce',
  [NEUTRAL_COLOUR_NAMES[3]]: '#949cbb',
  [NEUTRAL_COLOUR_NAMES[4]]: '#838ba7',
  [NEUTRAL_COLOUR_NAMES[5]]: '#737994',
  [NEUTRAL_COLOUR_NAMES[6]]: '#626880',
  [NEUTRAL_COLOUR_NAMES[7]]: '#51576d',
  [NEUTRAL_COLOUR_NAMES[8]]: '#414559',
  [NEUTRAL_COLOUR_NAMES[9]]: '#303446',
  [NEUTRAL_COLOUR_NAMES[10]]: '#292c3c',
  [NEUTRAL_COLOUR_NAMES[11]]: '#232634',
} as const satisfies FlavourPalette;
export const COLOURS_FLAVOUR_MACCHIATO = {
  [ACCENT_COLOUR_NAMES[0]]: '#f4dbd6',
  [ACCENT_COLOUR_NAMES[1]]: '#f0c6c6',
  [ACCENT_COLOUR_NAMES[2]]: '#f5bde6',
  [ACCENT_COLOUR_NAMES[3]]: '#c6a0f6',
  [ACCENT_COLOUR_NAMES[4]]: '#ed8796',
  [ACCENT_COLOUR_NAMES[5]]: '#ee99a0',
  [ACCENT_COLOUR_NAMES[6]]: '#f5a97f',
  [ACCENT_COLOUR_NAMES[7]]: '#eed49f',
  [ACCENT_COLOUR_NAMES[8]]: '#a6da95',
  [ACCENT_COLOUR_NAMES[9]]: '#8bd5ca',
  [ACCENT_COLOUR_NAMES[10]]: '#91d7e3',
  [ACCENT_COLOUR_NAMES[11]]: '#7dc4e4',
  [ACCENT_COLOUR_NAMES[12]]: '#8aadf4',
  [ACCENT_COLOUR_NAMES[13]]: '#b7bdf8',
  [NEUTRAL_COLOUR_NAMES[0]]: '#cad3f5',
  [NEUTRAL_COLOUR_NAMES[1]]: '#b8c0e0',
  [NEUTRAL_COLOUR_NAMES[2]]: '#a5adcb',
  [NEUTRAL_COLOUR_NAMES[3]]: '#939ab7',
  [NEUTRAL_COLOUR_NAMES[4]]: '#8087a2',
  [NEUTRAL_COLOUR_NAMES[5]]: '#6e738d',
  [NEUTRAL_COLOUR_NAMES[6]]: '#5b6078',
  [NEUTRAL_COLOUR_NAMES[7]]: '#494d64',
  [NEUTRAL_COLOUR_NAMES[8]]: '#363a4f',
  [NEUTRAL_COLOUR_NAMES[9]]: '#24273a',
  [NEUTRAL_COLOUR_NAMES[10]]: '#1e2030',
  [NEUTRAL_COLOUR_NAMES[11]]: '#181926',
} as const satisfies FlavourPalette;
export const COLOURS_FLAVOUR_MOCHA = {
  [ACCENT_COLOUR_NAMES[0]]: '#f5e0dc',
  [ACCENT_COLOUR_NAMES[1]]: '#f2cdcd',
  [ACCENT_COLOUR_NAMES[2]]: '#f5c2e7',
  [ACCENT_COLOUR_NAMES[3]]: '#cba6f7',
  [ACCENT_COLOUR_NAMES[4]]: '#f38ba8',
  [ACCENT_COLOUR_NAMES[5]]: '#eba0ac',
  [ACCENT_COLOUR_NAMES[6]]: '#fab387',
  [ACCENT_COLOUR_NAMES[7]]: '#f9e2af',
  [ACCENT_COLOUR_NAMES[8]]: '#a6e3a1',
  [ACCENT_COLOUR_NAMES[9]]: '#94e2d5',
  [ACCENT_COLOUR_NAMES[10]]: '#89dceb',
  [ACCENT_COLOUR_NAMES[11]]: '#74c7ec',
  [ACCENT_COLOUR_NAMES[12]]: '#89b4fa',
  [ACCENT_COLOUR_NAMES[13]]: '#b4befe',
  [NEUTRAL_COLOUR_NAMES[0]]: '#cdd6f4',
  [NEUTRAL_COLOUR_NAMES[1]]: '#bac2de',
  [NEUTRAL_COLOUR_NAMES[2]]: '#a6adc8',
  [NEUTRAL_COLOUR_NAMES[3]]: '#9399b2',
  [NEUTRAL_COLOUR_NAMES[4]]: '#7f849c',
  [NEUTRAL_COLOUR_NAMES[5]]: '#6c7086',
  [NEUTRAL_COLOUR_NAMES[6]]: '#585b70',
  [NEUTRAL_COLOUR_NAMES[7]]: '#45475a',
  [NEUTRAL_COLOUR_NAMES[8]]: '#313244',
  [NEUTRAL_COLOUR_NAMES[9]]: '#1e1e2e',
  [NEUTRAL_COLOUR_NAMES[10]]: '#181825',
  [NEUTRAL_COLOUR_NAMES[11]]: '#11111b',
} as const satisfies FlavourPalette;

export const COLOURS = {
  [FLAVOUR_NAMES[0]]: COLOURS_FLAVOUR_LATTE,
  [FLAVOUR_NAMES[1]]: COLOURS_FLAVOUR_FRAPPE,
  [FLAVOUR_NAMES[2]]: COLOURS_FLAVOUR_MACCHIATO,
  [FLAVOUR_NAMES[3]]: COLOURS_FLAVOUR_MOCHA,
} as const satisfies ColourPalette;
