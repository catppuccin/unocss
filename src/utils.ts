export const hexToRGBA = (hex: string, opacity = 100) => {
  if (hex.startsWith('#')) hex = hex.slice(1);
  if (opacity > 100) opacity = 100;

  const split = hex.match(/.{1,2}/g)!;

  return [
    parseInt(split[0], 16),
    parseInt(split[1], 16),
    parseInt(split[2], 16),
    opacity / 100,
  ];
};
