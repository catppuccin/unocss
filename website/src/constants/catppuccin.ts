// TODO: Use the exported `palette` from the preset.
export const catppuccinFlavours = [
  'latte',
  'frappe',
  'macchiato',
  'mocha',
] as const;

export const catppuccinAccentColours = [
  'rosewater',
  'flamingo',
  'pink',
  'mauve',
  'red',
  'maroon',
  'peach',
  'yellow',
  'green',
  'teal',
  'sky',
  'sapphire',
  'blue',
  'lavender',
] as const;

export const catppuccinBaseColours = [
  'text',
  'subtext1',
  'subtext0',
  'overlay2',
  'overlay1',
  'overlay0',
  'surface2',
  'surface1',
  'surface0',
  'base',
  'mantle',
  'crust',
] as const;

export type CatppuccinFlavour = (typeof catppuccinFlavours)[number];
export type CatppuccinAccentColour = (typeof catppuccinAccentColours)[number];
export type CatppuccinBaseColour = (typeof catppuccinBaseColours)[number];
