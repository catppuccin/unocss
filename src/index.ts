import { _extendTheme } from './extend.ts';

import type { Preset } from '@unocss/core';

import type { UnoCSSCatppuccinOptions } from './types.ts';

/**
 * Catppuccin preset for UnoCSS.
 *
 * @param options - Options of the preset.
 */
<<<<<<< HEAD
export const presetCatppuccin = (options: UnoCSSCatppuccinOptions = {}) => {
=======
export default (options: UnoCSSCatppuccinOptions = {}) => {
>>>>>>> f3e6c09 (Export preset as `default`)
  const preset: Preset = { name: 'unocss-catppuccin' };

  const {
    mode = 'extend',
    themeKey = 'colors',
    prefix = 'ctp',
    defaultFlavour,
  } = options;

  switch (mode) {
    case 'extend':
      preset.extendTheme = _extendTheme({ prefix, themeKey, defaultFlavour });
      break;
    default:
      // TODO: Custom logging?
      throw new Error(`Unsupported mode provided: '${mode}'`);
  }

  return preset;
};
