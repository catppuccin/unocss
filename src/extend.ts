import { flavorEntries, flavors } from '@catppuccin/palette';

import type { CatppuccinFlavor } from '@catppuccin/palette';

import type { ExtendOptions } from './types';

/**
 * Extend theme to UnoCSS by passing this to `extendTheme` function.
 *
 * @param options - Options for extending the theme.
 * @returns The main function that supposed to be passed
 * to UnoCSS `extendTheme` option.
 */
export const _extendTheme = (options: ExtendOptions = {}) => {
  const { themeKey = 'colors', prefix = 'ctp', defaultFlavour } = options;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (theme: Record<string, any>) => {
    theme[themeKey] ??= {};

    const targetObject =
      prefix ? (theme[themeKey][prefix] ??= {}) : theme[themeKey];

    if (defaultFlavour && flavors[defaultFlavour]) {
      addColoursToTarget(
        targetObject,
        flavors[defaultFlavour].colorEntries,
        prefix,
      );
    } else {
      for (const [flavourName, flavour] of flavorEntries) {
        targetObject[flavourName] ??= {};
        addColoursToTarget(
          targetObject[flavourName],
          flavour.colorEntries,
          prefix,
        );
      }
    }
  };
};

function addColoursToTarget(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  targetObject: Record<string, any>,
  colourEntries: CatppuccinFlavor['colorEntries'],
  prefix: string | false,
) {
  for (const [colourName, colour] of colourEntries) {
    if (!prefix && targetObject[colourName]) {
      targetObject['ctp'] ??= {};
      targetObject['ctp'][colourName] = colour.hex;
    } else {
      targetObject[colourName] = colour.hex;
    }
  }
}
