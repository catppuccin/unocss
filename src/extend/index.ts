// TODO: Get rid of `any` usages.
import { FLAVOURS, type PaletteColours } from '../palette.ts';

import type { ExtendOptions } from './types.ts';

/**
 * Extend theme to UnoCSS by passing this to `extendTheme` function.
 *
 * @param options - Options for extending the theme.
 * @returns The main function that supposed to be passed
 * to UnoCSS `extendTheme` option.
 */
export const _extendTheme = (options: ExtendOptions = {}) => {
  const { themeKey = 'colors', prefix = 'ctp', defaultFlavour } = options;

  const addFlavourColours = (
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    targetObj: Record<string, any>,
    flavour: PaletteColours,
    namespace?: string,
  ) => {
    if (namespace) {
      if (
        targetObj[namespace] !== undefined
        && typeof targetObj[namespace] !== 'object'
      ) {
        targetObj['ctp'] ??= {};
        targetObj['ctp'][namespace] ??= {};
        Object.assign(targetObj['ctp'][namespace], flavour);
        return;
      }

      targetObj[namespace] ??= {};
      Object.assign(targetObj[namespace], flavour);
      return;
    }

    if (!prefix) {
      targetObj['ctp'] ??= {};
      const ctpObj = targetObj['ctp']; // Cache this reference

      for (const [colorId, color] of Object.entries(flavour)) {
        if (colorId in targetObj) {
          // 'in' operator is slightly faster
          ctpObj[colorId] = color;
        } else {
          targetObj[colorId] = color;
        }
      }
    } else {
      Object.assign(targetObj, flavour);
    }
  };

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return (theme: Record<string, any>) => {
    theme[themeKey] ??= {};
    const targetObject =
      prefix ? (theme[themeKey][prefix] ??= {}) : theme[themeKey];

    if (defaultFlavour && defaultFlavour in FLAVOURS) {
      addFlavourColours(targetObject, FLAVOURS[defaultFlavour]);
    } else {
      for (const [flavourIdentifier, flavour] of Object.entries(FLAVOURS)) {
        addFlavourColours(targetObject, flavour, flavourIdentifier);
      }
    }
  };
};
