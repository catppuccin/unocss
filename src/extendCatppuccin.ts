import {
  variants as ctpVariants,
  labels as ctpLabels,
} from '@catppuccin/palette';
import { capitalizeChar } from './utils';

import type { ExtenderOptions } from './types';

export const extendCatppuccin = (options?: ExtenderOptions) => {
  const prefixOption = options?.prefix;
  const variantOption = options?.variant;

  const coloursObject = {};
  if (variantOption && ctpVariants[variantOption]) {
    for (const label of Object.keys(ctpLabels)) {
      const key = prefixOption
        ? `${prefixOption}${capitalizeChar(label)}`
        : label;

      coloursObject[key] = ctpLabels[label][variantOption].hex;
    }
  } else {
    for (const variant of Object.keys(ctpVariants)) {
      for (const label of Object.keys(ctpLabels)) {
        const key = prefixOption
          ? `${prefixOption}${capitalizeChar([variant, label]).join('')}`
          : `${variant}${capitalizeChar(label)}`;

        coloursObject[key] = ctpVariants[variant][label].hex;
      }
    }
  }

  return coloursObject;
};
