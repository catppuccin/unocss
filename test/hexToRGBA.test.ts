import { expect, test } from 'vitest';
import catppuccin from '@catppuccin/palette';

import { hexToRGBA } from '../src/utils';

test('HEX to RGBA', () => {
  const colors = [];
  Object.keys(catppuccin.variants).forEach((variant) => {
    colors.push(hexToRGBA(catppuccin.variants[variant].base.hex));
  });

  expect(colors).toMatchSnapshot();
});

test('HEX to RGBA with custom opacity', () => {
  const opacities = [15, 99, 1, 1709];
  const colors = [];
  Object.keys(catppuccin.variants).forEach((variant, index) => {
    colors.push(
      hexToRGBA(catppuccin.variants[variant].pink.hex, opacities[index])
    );
  });

  expect(colors).toMatchSnapshot();
});
