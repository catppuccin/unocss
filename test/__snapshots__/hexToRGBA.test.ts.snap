// Vitest Snapshot v1

exports[`HEX to RGBA 1`] = `
[
  [
    239,
    241,
    245,
    1,
  ],
  [
    48,
    52,
    70,
    1,
  ],
  [
    36,
    39,
    58,
    1,
  ],
  [
    30,
    30,
    46,
    1,
  ],
]
`;

exports[`HEX to RGBA with custom opacity 1`] = `
[
  [
    234,
    118,
    203,
    0.15,
  ],
  [
    244,
    184,
    228,
    0.99,
  ],
  [
    245,
    189,
    230,
    0.01,
  ],
  [
    245,
    194,
    231,
    1,
  ],
]
`;
